instrumentation/apache-camel-2.20/javaagent/src/test/groovy/test/SqsCamelTest.groovy

```diff
@@ -0,0 +1,183 @@
+    if (server != null) {
+      server.close()
+      server = null
+    }
+    if (sqs != null) {
+      sqs.stopAndWait()
+    }
+  }
+
+  def "simple sqs producer-consumer services"() {
+    setup:
+    def camelContext = server.getBean(CamelContext)
+    ProducerTemplate template = camelContext.createProducerTemplate()
+
+    when:
+    template.sendBody("direct:input", "{\"type\": \"hello\"}")
+    // wait for SQS to send back
+    Thread.sleep(2000)
+
+    then:
+    assertTraces(5) {
+      trace(0, 3) {
+        def parent = it
+        it.span(0) {
+          name "input"
+          kind INTERNAL
+          hasNoParent()
+          attributes {
+            "apache-camel.uri" "direct://input"
+          }
+        }
+        it.span(1) {
+          name "sqsCamelTest"
+          kind PRODUCER
+          parentSpanId parent.span(0).spanId
+          attributes {
+            "apache-camel.uri" "aws-sqs://sqsCamelTest?amazonSQSClient=%23sqsClient"
+            "messaging.destination" "sqsCamelTest"
+          }
+        }
+        it.span(2) {
+          name "sqsCamelTest"
+          kind CONSUMER
+          parentSpanId parent.span(1).spanId
+          attributes {
+            "apache-camel.uri" "aws-sqs://sqsCamelTest?amazonSQSClient=%23sqsClient&messageAttributeNames=traceparent"
+            "messaging.destination" "sqsCamelTest"
+            "messaging.message_id" String
+          }
+        }
+      }
+      trace(1, 1) {
+        it.span(0) {
+          name "SQS.ReceiveMessage"
+          kind CLIENT
+          hasNoParent()
+          attributes {
+            "aws.agent" "java-aws-sdk"
+            "aws.endpoint" "http://localhost:$sqsPort"
+            "aws.operation" "ReceiveMessageRequest"
+            "aws.queue.url" "http://localhost:$sqsPort/queue/sqsCamelTest"
+            "aws.service" "AmazonSQS"
+            "http.flavor" "1.1"
+            "http.method" "POST"
+            "http.status_code" 200
+            "http.url" "http://localhost:$sqsPort"
+            "net.peer.name" "localhost"
+            "net.peer.port" sqsPort
+            "net.transport" "IP.TCP"
+          }
+        }
+      }
+      trace(2, 1) {
+        it.span(0) {
+          name "SQS.DeleteMessage"
+          kind CLIENT
+          hasNoParent()
+          attributes {
+            "aws.agent" "java-aws-sdk"
+            "aws.endpoint" "http://localhost:$sqsPort"
+            "aws.operation" "DeleteMessageRequest"
+            "aws.queue.url" "http://localhost:$sqsPort/queue/sqsCamelTest"
+            "aws.service" "AmazonSQS"
+            "http.flavor" "1.1"
+            "http.method" "POST"
+            "http.status_code" 200
+            "http.url" "http://localhost:$sqsPort"
+            "net.peer.name" "localhost"
+            "net.peer.port" sqsPort
+            "net.transport" "IP.TCP"
+          }
+        }
+      }
+      trace(3, 1) {
+        it.span(0) {
+          name "SQS.ReceiveMessage"
+          kind CLIENT
```

Comment:
Just a question: can you explain those separate AWS CLIENT traces? I'm interested in a reason why those spans appear separately

Reply:
Delete is easy - it's just "cleanup" done after processing batch, so after produce / consume trace is done.

Reply:
"Receive" spans are quite a different story - it took me while to understand whole process :D 

Basically, when SQS consumer (bean) is up, it starts sending "Receive" messages to SQS. Each such message can be seen as a CONSUMER, but only after a message is in fact available and consumed. So, once one of the messages succeeds, Camel starts "consumer" route and that's where Camel's instrumentation kicks in, creating a CONSUMER span. As there is a propagation enabled for Camel, the CONSUMER is tied to Camel's PRODUCER span issued earlier. So far so good. AWS SDK instrumentation does not have a trace propagation (!), therefore "RECEIVE" span will hang detached from the internal-producer-consumer trace.

Now things are getting interesting. Once I enable trace propagation for AWS SDK (and change span KIND for SQS "ReceiveMessage" which IMO should be CONSUMER not CLIENT), there are two options:
- two CONSUMER spans in the trace: internal - producer (camel) - consumer (AWS SDK) - consumer (camel)
- camel consumer will be suppressed (!) and the trace will have one consumer span - the one from AWS SDK: internal - producer (camel) - consumer (AWS SDK)

@anuraaga @trask @jkwatson I'd very much like to hear your opinions on the above question (what should happen when there is Camel and AWS SDK with trace propagation) ?

By the way - thanks @mateuszrzeszutek for bringing up this matter :)

Reply:
I think it'd make the most sense if camel spans were internal - in the context of this test:

```
INTERNAL input (camel)
--> INTERNAL sqsCamelTest (camel)
    --> PRODUCER SQS.SendMessage (AWS SDK)
        --> CONSUMER SQS.ReceiveMessage (AWS SDK)
            --> INTERNAL sqsCamelTest (camel)
```

Camel is just a wrapper library that does not send any message to any queue directly, it relies on other libraries - and it makes sense for the camel instrumentation to rely on other instrumentations (AWS SDK here).

We need to have trace propagation in AWS SDK and change SQS spans to PRODUCER/CONSUMER for it to happen though.